Write vendor, regardless of lock, if force is true

carolynvs opened the pull request:
I wasn't taking forcevendor into account properly when the diffs were the same (which is exactly what that flag was for...) 😞 

This fixes my misunderstanding about when lock and newlock should be set in the SafeWriter tests.  Previously, I thought I should only pass in newlock when the lock had changed, or was brand new, when actually it should be passed anytime there was an existing lock file. Due to the bad test, it was missing that a bare ensure with an existing lock file (and no vendor) should result in vendor being written.

This fixes #323.

jstemmer commented on txn_writer.go:
```diff
@@ -162,20 +159,23 @@ func (diff StringDiff) MarshalJSON() ([]byte, error) {
-			sw.Payload.ForceWriteVendor = true
+			sw.Payload.WriteVendor = true
 		} else {
 			diff := diffLocks(lock, newLock)
```
Running `dep ensure -n` without a `lock.json` file does not print a diff even though `dep ensure` creates a new lock file. I don't think that's WAI?

With the other changes you made here this entire block in `newLock != nil` could then be simplified to:
```go
if newLock != nil {
    sw.Payload.Lock = newLock
    sw.Payload.LockDiff = diffLocks(lock, newLock)
    sw.Payload.WriteVendor = true
} ...
```
which solves the problem. However, you'd probably have to fix the [nil check](https://github.com/golang/dep/blob/55422bc13ba89f99c3889ccdca701799f2d1d6a0/txn_writer.go#L390) in the diffLocks function for this to work when lock is nil.

carolynvs replied:
Thanks! I'll play with that simplification and add a test for a bare ensure without an existing lock file.

carolynvs commented:
* I've fixed ensure dry-run to print the entire lock file when it is brand new, only showing the diff when there was an existing lock.
* I've simplified `SafeWriter.Prepare`.

carolynvs commented:
This is ready for review

jstemmer commented on txn_writer.go:
```diff
@@ -160,26 +157,23 @@ func (diff StringDiff) MarshalJSON() ([]byte, error) {
 // - If the forceVendor param is true, then vendor/ will be unconditionally
 //   written out based on newLock if present, else lock, else error.
-func (sw *SafeWriter) Prepare(manifest *Manifest, lock *Lock, newLock *Lock, forceVendor bool) {
+func (sw *SafeWriter) Prepare(manifest *Manifest, oldLock *Lock, newLock *Lock, forceVendor bool) {
```
nit: The function comments still refer to `lock`

sdboyer replied:
Also, s/`oldLock *Lock, newLock *Lock`/`oldLock, newLock *Lock`)

carolynvs replied:
Doh! 😊 Fixing...

sdboyer commented:
This is looking pretty good! I want to do a more thorough pass a little later, but aside from that nit, it seems like this should be able to go in presently. Thanks!

carolynvs commented:
I'm reviewing the original doc on that function (from before I ever touched it) and am a bit unsure about this scenario: 

`If sw.Lock is provided without an sw.NewLock, it will be written to the standard lock file name in the root dir, but vendor will NOT be written`

I believe the only usage of this function where the original lock is nil and the new lock is specified is  during `dep init` but that is always passing `forceVendor=true`. Is this still a valid scenario (no existing lock, brand new lock, don't write to vendor)? If so, then I need to add a test for that and fix my implementation.

carolynvs commented:
I've updated the lock diff to include the memo, e.g. `Memo: abc123 -> def546`.

sdboyer commented:
I think the case I was aiming for with that original doc was a bit of forward-thinking - that we'd want to `dep ensure -no-vendor`, where we ensure the correct lock is written, but skip writing out vendor.

We're not quite ready for that now, I think, but I'd like to see the safe writer preserve support for such a case when we get to it.

carolynvs commented:
Gotcha! I'll get that fixed tonight.

carolynvs commented:
Apologies for the delay. It was really bothering me how much logic was being "encoded" in the combination of function arguments and I wanted to make this more explicit/maintainable for the next person.

* If `manifest` is provided, write it.
* If `newLock` is provided, write it.
* Control whether vendor/ is written using: `VendorOnChanged`, `VendorAlways`, `VendorNever`.
* Write vendor/ using `newLock`.

So for the future scenario of `dep ensure --no-vendor`, the caller passes `VendorNever` instead of varying whether oldLock or newLock is set.

sdboyer commented:
This looks to be in pretty good shape, so I'm gonna merge - thanks!
